example-application/libraries/message-queue-client.js

```diff
@@ -16,17 +16,52 @@ class MessageQueueClient extends EventEmitter {
     // It can get one in the constructor here or even change by environment variables
     if (customMessageQueueProvider) {
       this.messageQueueProvider = customMessageQueueProvider;
-    } 
-    else if(process.env.USE_FAKE_MQ === 'true'){
+    } else if (process.env.USE_FAKE_MQ === 'true') {
       this.messageQueueProvider = new FakeMessageQueueProvider();
-    }
-    else {
+    } else {
       this.messageQueueProvider = new FakeMessageQueueProvider();
     }
 
     this.countEvents();
   }
 
+  // This function stores all the MQ events in a local data structure so later
+  // one query this
+  countEvents() {
+    if (this.recordingStarted === true) {
+      return; // Already initialized and set up
+    }
+    this.records = {
+      ack: { count: 0, lastEventData: null },
+      nack: { count: 0, lastEventData: null },
+      publish: { count: 0, lastEventData: null },
+      consume: { count: 0, lastEventData: null },
+    };
+    ['nack', 'ack', 'publish'].forEach((eventToListenTo) => {
+      this.on(eventToListenTo, (eventData) => {
+        this.records[eventToListenTo].count++;
+        this.records[eventToListenTo].lastEventData = eventData;
+        this.emit('message-queue-event', {
+          name: eventToListenTo,
+          eventsRecorder: this.recordingStarted,
+        });
+      });
+    });
+  }
+
+  // Helper methods for testing - Resolves/fires when some event happens
+  async waitFor(eventName, howMuch) {
+    return new Promise((resolve, reject) => {
+      // The first resolve is for cases where the caller has approached AFTER the event has already happen
+      this.resolveIfEventExceededThreshold(eventName, howMuch, resolve);
+      this.on('message-queue-event', (eventInfo) => {
+        this.resolveIfEventExceededThreshold(eventName, howMuch, resolve);
+      });
```

Comment:
I suggest converting to this so it will clean the resources (unless you wanna show it in the presentation and it gonna make the code larger 😬)

```suggestion
      // The first resolve is for cases where the caller has approached AFTER the event has already happen
      if (this.resolveIfEventExceededThreshold(eventName, options, resolve)) {
        return;
      }

      const handler = (eventInfo) => {
        if (this.resolveIfEventExceededThreshold(eventName, options, resolve)) {
          this.off('message-queue-event', handler);
        }
      };

      this.on('message-queue-event', handler);
```